contrib/go/src/python/pants/contrib/go/tasks/go_protobuf_gen.py

```diff
@@ -0,0 +1,127 @@
+    return self._protoc.select(context=self.context)
+
+  @property
+  def _version(self):
+    return self._protoc.version(context=self.context)
+
+  @memoized_property
+  def _protoc(self):
+    return Protoc.scoped_instance(self)
+
+  @memoized_property
+  def _protoc_gen_go(self):
+    return ProtocGenGo.scoped_instance(self)
+
+  def synthetic_target_extra_dependencies(self, target, target_workdir):
+    import_target = self.get_options().import_target
+    if import_target is None:
+      raise TaskError('Option import_target in scope {} must be set.'.format(
+        self.options_scope))
+    return self.context.resolve(import_target)
+
+  def synthetic_target_type(self, target):
+    return GoProtobufGenLibrary
+
+  def is_gentarget(self, target):
+    return isinstance(target, GoProtobufLibrary)
+
+  @classmethod
+  def product_types(cls):
+    return ['go']
+
+  def execute_codegen(self, target, target_workdir):
+    target_cmd = [self._binary]
+
+    env = os.environ.copy()
+    env['PATH'] = ':'.join([env['PATH'],
+                            os.path.dirname(self._protoc_gen_go.select(context=self.context))])
+
+    bases = OrderedSet(tgt.target_base for tgt in target.closure() if self.is_gentarget(tgt))
+    for base in bases:
+      target_cmd.append('-I={}'.format(os.path.join(get_buildroot(), base)))
+
+    outdir = os.path.join(target_workdir, 'src', 'go')
+    safe_mkdir(outdir)
+    target_cmd.append('--go_out={}'.format(outdir))
+
+    all_sources = list(target.sources_relative_to_buildroot())
+    for source in all_sources:
+      file_cmd = target_cmd + [os.path.join(get_buildroot(), source)]
+      with self.context.new_workunit(name=source,
+                                     labels=[WorkUnitLabel.TOOL],
+                                     cmd=' '.join(file_cmd)) as workunit:
+        self.context.log.info(' '.join(file_cmd))
+        result = subprocess.call(file_cmd,
+                                 env=env,
+                                 stdout=workunit.output('stdout'),
+                                 stderr=workunit.output('stderr'))
+        if result != 0:
+          raise TaskError('{} ... exited non-zero ({})'.format(self._binary, result))
+
+  @property
+  def _copy_target_attributes(self):
+    return [a for a in super(GoProtobufGen, self)._copy_target_attributes if a != 'provides']
+
+  def synthetic_target_dir(self, target, target_workdir):
+    all_sources = list(target.sources_relative_to_buildroot())
+    source = all_sources[0]
+    namespace = self._get_go_namespace(source)
+    return os.path.join(target_workdir, 'src', 'go', namespace.replace(".", os.path.sep))
+
+  @classmethod
+  def _get_go_namespace(cls, source):
+    with open(source) as fh:
+      namespace = cls._NAMESPACE_PARSER.search(fh.read())
+      if not namespace:
+        raise TaskError('File {} must contain "option go_package"'.format(source))
+      return namespace.group(1)
+
+  # Override because file names were too long.
```

Comment:
Could you open a ticket about this one? Or fix it in the base class maybe?

Reply:
Here's an example of the error:

```
Exception caught: (<type 'exceptions.IOError'>)
  File "/Users/travis/src/pants/src/python/pants/bin/pants_loader.py", line 73, in <module>
    main()
  File "/Users/travis/src/pants/src/python/pants/bin/pants_loader.py", line 69, in main
    PantsLoader.run()
  File "/Users/travis/src/pants/src/python/pants/bin/pants_loader.py", line 65, in run
    cls.load_and_execute(entrypoint)
  File "/Users/travis/src/pants/src/python/pants/bin/pants_loader.py", line 58, in load_and_execute
    entrypoint_main()
  File "/Users/travis/src/pants/src/python/pants/bin/pants_exe.py", line 39, in main
    PantsRunner(exiter, start_time=start_time).run()
  File "/Users/travis/src/pants/src/python/pants/bin/pants_runner.py", line 53, in run
    return runner.run()
  File "/Users/travis/src/pants/src/python/pants/bin/local_pants_runner.py", line 45, in run
    self._run()
  File "/Users/travis/src/pants/src/python/pants/bin/local_pants_runner.py", line 90, in _run
    goal_runner_result = goal_runner.run()
  File "/Users/travis/src/pants/src/python/pants/bin/goal_runner.py", line 272, in run
    result = self._execute_engine()
  File "/Users/travis/src/pants/src/python/pants/bin/goal_runner.py", line 260, in _execute_engine
    result = engine.execute(self._context, self._goals)
  File "/Users/travis/src/pants/src/python/pants/engine/legacy_engine.py", line 26, in execute
    self.attempt(context, goals)
  File "/Users/travis/src/pants/src/python/pants/engine/round_engine.py", line 233, in attempt
    goal_executor.attempt(explain)
  File "/Users/travis/src/pants/src/python/pants/engine/round_engine.py", line 49, in attempt
    task.execute()
  File "/Users/travis/src/pants/contrib/go/src/python/pants/contrib/go/tasks/go_compile.py", line 66, in execute
    self._go_install(vt.target, gopath, build_flags)
  File "/Users/travis/src/pants/contrib/go/src/python/pants/contrib/go/tasks/go_compile.py", line 107, in _go_install
    workunit_labels=[WorkUnitLabel.COMPILER])
  File "/Users/travis/src/pants/contrib/go/src/python/pants/contrib/go/subsystems/go_distribution.py", line 139, in execute_go_cmd
    stdout=workunit.output('stdout'),
  File "/Users/travis/src/pants/src/python/pants/base/workunit.py", line 193, in output
    self._outputs[name] = FileBackedRWBuf(path)
  File "/Users/travis/src/pants/src/python/pants/util/rwbuf.py", line 76, in __init__
    _RWBuf.__init__(self, open(backing_file, 'a+'))

Exception message: [Errno 63] File name too long: u'/Users/travis/src/pants/.pants.d/run-tracker/pants_run_2018_05_23_17_04_36_657_cf1fd75f8bd547b1bd06cce8dcf5ad19/tool_outputs/install__pants_d_gen_go_protobuf_252d64521cf9_contrib_go_examples_src_protobuf_org_pantsbuild_example_distance_distance_go_current_src_go_pantsbuild_example_distance_contrib_go_examples_src_protobuf_org_pantsbuild_example_distance_distance_go-d58513c8-2091-49eb-8339-b9ff1d96a514.stdout'
```

I made the `target.id` -> `target.name` change in the base class, though that affects other code generators too.

Reply:
Oooh.

So, in that case, I think that's because the name of the workunit is too long in `compile.go`, here: https://github.com/pantsbuild/pants/blob/master/contrib/go/src/python/pants/contrib/go/tasks/go_compile.py#L106 ... moving that to something shorter (or summarizing into a package name) might help.